### PR TITLE
MINIFICPP-1952 InvokeHTTP POST/PUT should support large files

### PR DESCRIPTION
**Pull request by fgerlits**
https://issues.apache.org/jira/browse/MINIFICPP-1952

When doing POST/PUT in InvokeHTTP, read the flow file contents in chunks during the upload instead of reading the whole thing at once into memory.

The change to `catch.hpp` is not part of this pull request, but of #1441.  That PR should be merged first, after which the `catch.hpp` change will disappear from this PR.

---

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/src/utils/BaseHTTPClient.cpp [hunk 1]
```diff
@@ -160,4 +160,118 @@ std::string URL::toString() const {
...
+  }
+  return 0;
+}
+
+size_t HTTPUploadByteArrayInputCallback::setPosition(int64_t offset) {
+  if (stop) {
+    return HTTPRequestResponse::SEEKFUNC_FAIL;
+  }
+  if (getBufferSize() <= static_cast<size_t>(offset)) {
+    return HTTPRequestResponse::SEEKFUNC_FAIL;
```
Comment (adamdebreceni) [inline libminifi/src/utils/BaseHTTPClient.cpp:243]: I see that this is the same as the original version (so we shouldn't tinker with it in this PR), but do you think it is ok to fail a seek to the end of the buffer?

File: libminifi/src/utils/BaseHTTPClient.cpp [hunk 2]
```diff
@@ -160,4 +160,118 @@ std::string URL::toString() const {
...
+  pos = offset;
+  seek(pos);
+  return HTTPRequestResponse::SEEKFUNC_OK;
+}
+
+size_t HTTPUploadStreamContentsCallback::getDataChunk(char *data, size_t size) {
+  logger_->log_trace("HTTPUploadStreamContentsCallback is asked for up to %zu bytes", size);
+
+  std::vector<std::byte> buffer(size);
+  size_t num_read = input_stream_->read(buffer);
```
Comment (adamdebreceni) [inline libminifi/src/utils/BaseHTTPClient.cpp:254]: can't we read directly into `data`? might there be a requirement to not mess up `data` if we can't read the full size (or maybe on error)?

File: libminifi/src/utils/BaseHTTPClient.cpp [hunk 3]
```diff
@@ -160,4 +160,118 @@ std::string URL::toString() const {
...
+}
+
+size_t HTTPUploadStreamContentsCallback::setPosition(int64_t offset) {
+  if (offset == 0) {
+    logger_->log_debug("HTTPUploadStreamContentsCallback is ignoring request to rewind to the beginning");
+  } else {
+    logger_->log_warn("HTTPUploadStreamContentsCallback is ignoring request to seek to position %" PRId64, offset);
+  }
+
+  return HTTPRequestResponse::SEEKFUNC_OK;
```
Comment (adamdebreceni) [inline libminifi/src/utils/BaseHTTPClient.cpp:274]: some streams (like `FileStream`) might properly support `seek`s, could not calling `seek` cause problems? (corrupted data, failed request etc.) should we return `SEEKFUNC_FAIL` here?

File: libminifi/src/utils/BaseHTTPClient.cpp (hunk 2 above)
Comment (fgerlits) [reply libminifi/src/utils/BaseHTTPClient.cpp:254]: I don't know if there is such a requirement.  You're right, it's probably safe to read into the buffer directly.  I'll change it.

File: libminifi/src/utils/BaseHTTPClient.cpp (hunk 3 above)
Comment (fgerlits) [reply libminifi/src/utils/BaseHTTPClient.cpp:274]: Yeah, it's not clear to me when and why curl would call seek during an upload.  One thing I've seen during testing was a call to `seek(0)` _after_ the upload; if this call returns `SEEKFUNC_FAIL` then the upload fails, but if I don't do anything just return `SEEKFUNC_OK` then it works fine.

What do you think we should do?

File: libminifi/src/utils/BaseHTTPClient.cpp (hunk 1 above)
Comment (adamdebreceni) [reply libminifi/src/utils/BaseHTTPClient.cpp:243]: yes a Jira would be great

Comment (fgerlits) [reply libminifi/src/utils/BaseHTTPClient.cpp:243]: https://issues.apache.org/jira/browse/MINIFICPP-1971

File: libminifi/src/utils/BaseHTTPClient.cpp (hunk 3 above)
Comment (adamdebreceni) [reply libminifi/src/utils/BaseHTTPClient.cpp:272]: I see you have added it to this jira (https://issues.apache.org/jira/browse/MINIFICPP-1971), we will investigate there

**Review by adamdebreceni — APPROVED** [review]

**Review by szaszm — APPROVED** [review]

File: libminifi/src/utils/BaseHTTPClient.cpp [hunk 4]
```diff
@@ -160,4 +160,116 @@ std::string URL::toString() const {
...
+    size_t len = buffer_size - pos;
+    if (len <= 0) {
+      return 0;
+    }
+    auto *ptr = getBuffer(pos);
+
+    if (ptr == nullptr) {
+      return 0;
+    }
+    if (len > size)
```
Comment (adam-markovics) [inline libminifi/src/utils/BaseHTTPClient.cpp:228]: I would use `std::min` instead

File: libminifi/src/utils/BaseHTTPClient.cpp [hunk 5]
```diff
@@ -160,4 +160,116 @@ std::string URL::toString() const {
...
+}
+
+size_t HTTPUploadByteArrayInputCallback::getDataChunk(char *data, size_t size) {
+  if (stop) {
+    return HTTPRequestResponse::CALLBACK_ABORT;
+  }
+  size_t buffer_size = getBufferSize();
+  if (pos <= buffer_size) {
+    size_t len = buffer_size - pos;
+    if (len <= 0) {
```
[the remark this answers is not in the document]
Comment (fgerlits) [reply libminifi/src/utils/BaseHTTPClient.cpp:220]: This is pre-existing code, I'd rather not change it in this pull request.  I have added your suggestions to https://issues.apache.org/jira/browse/MINIFICPP-1971.

File: libminifi/src/utils/BaseHTTPClient.cpp (hunk 4 above)
Comment (fgerlits) [reply libminifi/src/utils/BaseHTTPClient.cpp:228]: Also added to https://issues.apache.org/jira/browse/MINIFICPP-1971.

**Review by adam-markovics — APPROVED** [review]

**Pull request closed without merging**